File: leetcode/python/2225_find_players_with_zero_or_one_losses.py
class Solution:
    def findWinners(self, matches: list[list[int]]) -> list[list[int]]:
        wins, lost1, loses = set[int](), set[int](), set[int]()
        for winner, loser in matches:
            if winner not in lost1 and winner not in loses:
                wins.add(winner)
            if loser in lost1:
                lost1.remove(loser)
                loses.add(loser)
            elif loser not in loses:
                lost1.add(loser)
                if loser in wins:
                    wins.remove(loser)
        return [sorted(wins), sorted(lost1)]
    
s = Solution()
print(s.findWinners([[1,3],[2,3],[3,6],[5,6],[5,7],[4,5],[4,8],[4,9],[10,4],[10,9]]))
print(s.findWinners([[2,3],[1,3],[5,4],[6,4]]))

File: leetcode/rust/src/2225_find_players_with_zero_or_one_losses.rs
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn find_winners(matches: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let mut lose_count = HashMap::new();
        let mut players = HashSet::new();

        for pair in matches {
            let winner = pair[0];
            let loser = pair[1];

            players.insert(winner);
            players.insert(loser);
            lose_count.entry(loser).and_modify(|f| *f += 1).or_insert(1);
        }

        let mut no_lose = Vec::new();
        let mut one_lose = Vec::new();
        let mut players = players.into_iter().collect::<Vec<i32>>();
        players.sort_unstable();
        for player in players {
            match lose_count.get(&player) {
                None => no_lose.push(player),
                Some(1) => one_lose.push(player),
                _ => (),
            }
        }
        vec![no_lose, one_lose]
    }
}